engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala

```diff
@@ -172,29 +172,18 @@ class DevelopmentDeploymentManager(actorSystem: ActorSystem)
   override def invokeCustomAction(
       actionRequest: CustomActionRequest,
       canonicalProcess: CanonicalProcess
-  ): Future[Either[CustomActionError, CustomActionResult]] =
-    Future.successful(
-      customActions
-        .find(_.name.equals(actionRequest.name))
-        .map(customAction => {
-          val processName = actionRequest.processVersion.processName
-          val statusDetails =
-            memory.getOrElse(processName, createAndSaveProcessState(NotDeployed, actionRequest.processVersion))
-
-          if (customAction.allowedStateStatusNames.contains(statusDetails.status.name)) {
-            customActionStatusMapping
-              .get(customAction)
-              .map { status =>
-                asyncChangeState(processName, status)
-                Right(CustomActionResult(actionRequest, s"Done ${actionRequest.name}"))
-              }
-              .getOrElse(Left(CustomActionInvalidStatus(actionRequest, statusDetails.status.name)))
-          } else {
-            Left(CustomActionInvalidStatus(actionRequest, statusDetails.status.name))
-          }
-        })
-        .getOrElse(Left(CustomActionNotImplemented(actionRequest)))
-    )
+  ): Future[CustomActionResult] = {
+    val processName = actionRequest.processVersion.processName
+    val statusOpt = customActionStatusMapping
+      .collectFirst { case (customAction, status) if customAction.name == actionRequest.name => status }
+
+    statusOpt match {
+      case Some(newStatus) =>
+        asyncChangeState(processName, newStatus)
+        Future.successful(CustomActionResult(actionRequest, s"Done ${actionRequest.name}"))
+      case _ => ???
```

Comment:
IMO it should be inside Future.apply - the whole block probably should be. Generally, it is not a good idea to throw exceptions inside a method that declares that returns the `Future`. Invoking service can handle wrong such a case.